ADR/ADR022-use-postgresql-for-form-creation-in-form-admin.md

```diff
@@ -0,0 +1,70 @@
+
+Currently we use the [form object pattern](https://dev.to/drbragg/rails-design-patterns-form-object-4d47#:~:text=Form%20Object%20Pattern,consistent%20API%20in%20our%20code.) to manage the intermediate state of pages as they are being created.
+
+ This gives each screen a [model](https://github.com/alphagov/forms-admin/tree/main/app/forms/pages), [controller](https://github.com/alphagov/forms-admin/tree/main/app/controllers/pages) and [view](https://github.com/alphagov/forms-admin/tree/main/app/views/pages).
+
+We store the state of incomplete `pages` in the session hash. The code updating this hash is spread across the form objects and controllers. When adding or modifying steps in the `page` journey, it's up to the programmer to make sure the session is updated correctly, such as making sure it's cleared in the right places.
+
+As we've added more features for `pages` such as extra input types, conditions and additional guidance, managing state in the session hash has become harder. We've had issues around producing inconsistent data, using a mix of strings and symbols and deleting data at the wrong point.
+
+Also, as the session hash is currently stored in user's cookie, there is a size limit of 4kb. This is not an immediate issue but could pose difficulties when a `page` has many question options, long guidance text or we start using the session in other parts of the application.
+
+We should:
+
+This ADR proposes that we stop using the forms object pattern to create pages. Instead each step is represented using an activeRecord model and database table. This means we no longer need to use the session hash to maintain the state of incomplete pages, removing the burden to manage it from the programmer. In essence we are moving from the "hard way" to the "easy way" in this [breakdown of multi-page forms](https://www.codewithjason.com/rails-multi-step-forms/). When a user finished configuring a `page` the different records would be collected together and submitted to the API as before.
+
+Essentially:
+
+* Refactor our models to be more Rails-standard and switch from ActiveModel (doesn't easily persist locally) to using ActiveRecord (allows us to easily persist locally) on our classes.
+
+* Rather than manually managing data in session for the user, as the user takes an action the data entered via the form is persisted to the PostgreSQL database for forms-admin.
+
+* Use the PostgreSQL database to manage data.
+
+* Drop using cookies for data in session (other than checking if the user is logged in)
+
+## Strategy
+
+* We add a `FormPages` class that is an ActiveRecord model. This is the join table between Forms and Pages in the database.
+* Temporarily, we keep use of `FormObjects` to break up the record into multiple pages and validate at those steps, before saving the record.
+
+## Decision
+
+> Accepted?
+
+## Consequences
```

Comment:
If we hypothetically wanted to monolith-ise our apps in future, do we think this change would make that easier, harder, or make no difference?

(my instinct is that this might make it slightly easier but you'll have a clearer idea than I do)

Reply:
It probably will be easier _if_ we decided to Monolithi-cise (but from pov I don't want to).